### PR TITLE
bwl: channel switch nl80211

### PR DESCRIPTION
**Pull request by YuriyMasechko**
Implement channel switch for nl80211.

Channel switch command for non intel hostapd doesn't support ACS.
Only command with given channel was implemented.
#366 

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

### CODE REVIEW

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 1]
```diff
@@ -92,6 +94,18 @@ static ap_wlan_hal::Event nl80211_to_bwl_event(const std::string &opcode)
...
+static uint8_t wpa_bw_to_beerocks_bw(const std::string &chan_width)
+{
+    // 20 MHz (no HT)
+    // 20 MHz
+    // 40 MHz
+    // 80 MHz
+    // 80+80 MHz
+    // 160 MHz
+
+    return chan_width == "80+80" ? 160 : beerocks::string_utils::stoi(chan_width);
```
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]:
```suggestion
    return (chan_width == "80+80") ? 160 : beerocks::string_utils::stoi(chan_width);
```

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 2]
```diff
@@ -614,8 +667,23 @@ bool ap_wlan_hal_nl80211::process_nl80211_event(parsed_obj_map_t &parsed_obj)
     } break;
 
     // ACS/CSA Completed
+    case Event::CTRL_EventChannelSwitch:
+        m_radio_info.channel =
+            beerocks::utils::wifi_freq_to_channel(beerocks::string_utils::stoi(parsed_obj["freq"]));
+        m_radio_info.bandwidth          = wpa_bw_to_beerocks_bw(parsed_obj["ch_width"]);
```
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: `wpa_bw_to_beerocks_bw` doesn't handle empty strings... Need to make sure that `parsed_obj["ch_width"]` is not empty.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 3]
```diff
@@ -281,12 +295,51 @@ bool ap_wlan_hal_nl80211::switch_channel(int chan, int bw, int vht_center_freque
     LOG(TRACE) << __func__ << " channel: " << chan << ", bw: " << bw
                << ", vht_center_frequency: " << vht_center_frequency;
 
-    // TODO: IMPLEMENT!
+    std::string cmd = "CHAN_SWITCH 5 ";
```
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:302]: What does the "5" means?
Probably should add a comment with an explanation of the command structure.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 4]
```diff
@@ -281,12 +295,51 @@ bool ap_wlan_hal_nl80211::switch_channel(int chan, int bw, int vht_center_freque
...
 
-    LOG(DEBUG) << "Got channel switch, simulate ACS-STARTED;ACS-COMPLETED";
-    event_queue_push(Event::ACS_Started);
-    event_queue_push(Event::ACS_Completed);
-    event_queue_push(Event::CSA_Finished);
+    if (chan == 0) {
+        LOG(ERROR) << "ACS i not supported";
```
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]:
```suggestion
        LOG(ERROR) << "ACS is not supported";
```

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 5]
```diff
@@ -281,12 +295,51 @@ bool ap_wlan_hal_nl80211::switch_channel(int chan, int bw, int vht_center_freque
...
+        return false;
+    } else {
+
+        std::string bandwidth_str =
+            std::to_string(beerocks::utils::convert_bandwidth_to_int((beerocks::eWiFiBandwidth)bw));
```
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: No need to store the string in a separate variable.
It can converted directly into the command string on line 328.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 6]
```diff
@@ -281,12 +295,51 @@ bool ap_wlan_hal_nl80211::switch_channel(int chan, int bw, int vht_center_freque
...
+        if (bw == beerocks::BANDWIDTH_20 || bw == beerocks::BANDWIDTH_40) {
+            cmd += " ht"; //n
+        } else if (bw == beerocks::BANDWIDTH_80) {
+            cmd += " vht"; // ac
+        }
+    }
+
+    // Send command
+    if (!wpa_ctrl_send_msg(cmd)) {
+        LOG(ERROR) << "switch_channel() failed!";
```
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]:
```suggestion
        LOG(ERROR) << "wpa_ctrl_send_msg() failed!";
```

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 7]
```diff
@@ -281,12 +295,51 @@ bool ap_wlan_hal_nl80211::switch_channel(int chan, int bw, int vht_center_freque
...
+        }
+
+        cmd += " bandwidth=" + bandwidth_str;
+
+        // Supported Standard n/ac
+        if (bw == beerocks::BANDWIDTH_20 || bw == beerocks::BANDWIDTH_40) {
+            cmd += " ht"; //n
+        } else if (bw == beerocks::BANDWIDTH_80) {
+            cmd += " vht"; // ac
+        }
```
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: What about 160Mhz?
```
enum eWiFiBandwidth : uint8_t {
    BANDWIDTH_20 = 0,
    BANDWIDTH_40,
    BANDWIDTH_80,
    BANDWIDTH_160,
    BANDWIDTH_MAX,
};
```

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 8]
```diff
@@ -614,8 +667,23 @@ bool ap_wlan_hal_nl80211::process_nl80211_event(parsed_obj_map_t &parsed_obj)
...
+        m_radio_info.vht_center_freq    = beerocks::string_utils::stoi(parsed_obj["cf1"]);
+        m_radio_info.is_dfs_channel     = beerocks::string_utils::stoi(parsed_obj["dfs"]);
+        m_radio_info.last_csa_sw_reason = ChanSwReason::Unknown;
+        if (son::wireless_utils::which_freq(m_radio_info.channel) == beerocks::eFreqType::FREQ_5G) {
+            m_radio_info.is_5ghz = true;
+        }
+        break;
     case Event::ACS_Completed:
     case Event::CSA_Finished:
+        if (event == Event::CSA_Finished)
```
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: It's kinda weird to have an `if` statement inside a `case` checking exactly the same value.
Perhaps it's better simply to break from `case Event::ACS_Completed` above...

**Review by vitalybu — CHANGES_REQUESTED** [review]

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 1 above)
Comment (morantr) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: I'm not that it is a good idea to unite "80+80" to "160". They are two different options.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 9]
```diff
@@ -281,12 +295,51 @@ bool ap_wlan_hal_nl80211::switch_channel(int chan, int bw, int vht_center_freque
...
+    std::string cmd = "CHAN_SWITCH 5 ";
 
-    LOG(DEBUG) << "Got channel switch, simulate ACS-STARTED;ACS-COMPLETED";
-    event_queue_push(Event::ACS_Started);
-    event_queue_push(Event::ACS_Completed);
-    event_queue_push(Event::CSA_Finished);
+    if (chan == 0) {
+        LOG(ERROR) << "ACS i not supported";
+        return false;
+    } else {
```
Comment (morantr) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: The `else` statement is redundant. You can leave the code without the `else`.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 5 above)
Comment (morantr) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: It is confusing that `bw` argument is an integer although it is treated as an enum. On dwpal implementation the int is not representing an enum, so here should be the same.

**Review by morantr — CHANGES_REQUESTED** [review]

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 5 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: There is a bug in dwpal,

in channel_selection_task:
channel_switch_request.bandwidth = beerocks::BANDWIDTH_80;

in ap_manager_thread print before switch call:
beerocks::utils::convert_bandwidth_to_int(
                          (beerocks::eWiFiBandwidth)request->cs_params().bandwidth)

So bw is passed as enum.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 1 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Done

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 3 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:302]: Comment added.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 4 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Done

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 5 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Done

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 6 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Done

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 7 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Added 160 Mhz for vht.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 8 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Done

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 1 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: There is no option for 80+80, there are already cases when 80+80 is used as 160.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 9 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Done

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 3 above)
Comment (vitalybu) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:302]: What does "cs_count" means and why is it hard-coded to 5?

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 2 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Done

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 3 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:302]: CSA_BCN_COUNT, it's beacon count before switch. 5 is default value in dwpal and default in hostapd_dfs_start_channel_switch()

@NL80211_CMD_CHANNEL_SWITCH: Perform a channel switch by announcing the
> + *	the new channel information (Channel Switch Announcement - CSA)
> + *	in the beacon for some time (as defined in the
> + *	%NL80211_ATTR_CH_SWITCH_COUNT parameter)

Comment (vitalybu) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:302]: Ok. Just add this to the comment and we're good ;)

**Review by vitalybu — APPROVED** [review]

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 3 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:302]: Done

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 5 above)
Comment (morantr) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: It seems to me like a bug in the ap_manager.
I mean, it is better to align with the integer representation rather than the enum.

Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: It's used as enum in all flow, starting from controller/cli/chan selection task.

File: common/beerocks/bwl/include/bwl/ap_wlan_hal.h [hunk 10]
```diff
@@ -53,6 +53,7 @@ class ap_wlan_hal : public virtual base_wlan_hal {
         ACS_Failed,
 
         CSA_Finished,
+        CTRL_EventChannelSwitch,
```
Comment (tomereli) [inline common/beerocks/bwl/include/bwl/ap_wlan_hal.h]: I think we can drop the prefix like other events anf call this `ChannelSwitch`

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 11]
```diff
@@ -614,8 +669,28 @@ bool ap_wlan_hal_nl80211::process_nl80211_event(parsed_obj_map_t &parsed_obj)
     } break;
 
     // ACS/CSA Completed
+    case Event::CTRL_EventChannelSwitch: {
```
Comment (tomereli) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: The comment should go to the next case

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp [hunk 12]
```diff
@@ -614,8 +669,28 @@ bool ap_wlan_hal_nl80211::process_nl80211_event(parsed_obj_map_t &parsed_obj)
...
+        m_radio_info.channel_ext_above  = beerocks::string_utils::stoi(parsed_obj["ch_offset"]);
+        m_radio_info.vht_center_freq    = beerocks::string_utils::stoi(parsed_obj["cf1"]);
+        m_radio_info.is_dfs_channel     = beerocks::string_utils::stoi(parsed_obj["dfs"]);
+        m_radio_info.last_csa_sw_reason = ChanSwReason::Unknown;
+        if (son::wireless_utils::which_freq(m_radio_info.channel) == beerocks::eFreqType::FREQ_5G) {
+            m_radio_info.is_5ghz = true;
+        }
+    } break;
     case Event::ACS_Completed:
+        break;
```
Comment (tomereli) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Why did you add a break here? It means we won't push an event for ACS completed.

**Review by tomereli — APPROVED** [review]
One major comment regarding not pushing event on ACS Done, otherwise LGTM.

**Review by morantr — APPROVED** [review]

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 12 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Done

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp (hunk 11 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp]: Done

File: common/beerocks/bwl/include/bwl/ap_wlan_hal.h (hunk 10 above)
Comment (YuriyMasechko) [reply common/beerocks/bwl/include/bwl/ap_wlan_hal.h]: Done
